addons/point_of_sale/static/src/app/store/pos_store.js

```diff
@@ -178,14 +176,142 @@ export class PosStore extends Reactive {
+
+            if (PaymentInterface) {
+                pm.payment_terminal = new PaymentInterface(this, pm);
+            }
+        }
+
+        // Create printer with hardware proxy, this will override related model data
+        this.unwatched.printers = [];
+        for (const relPrinter of this.models["pos.printer"].getAll()) {
+            const printer = relPrinter.serialize();
+            const HWPrinter = this.create_printer(printer);
+
+            HWPrinter.config = printer;
+            this.unwatched.printers.push(HWPrinter);
+
+            for (const id of printer.product_categories_ids) {
+                this.printers_category_ids_set.add(id);
+            }
+        }
+        this.config.iface_printers = !!this.unwatched.printers.length;
+
+        // Monitor product pricelist
+        this.computeProductPricelistCache();
+    }
+
+    computeProductPricelistCache() {
+        const date = DateTime.now();
+        const products = this.models["product.product"].getAll();
+        const pricelistItems = this.models["product.pricelist.item"].getAll();
+
+        for (const product of products) {
+            const applicableRules = {};
+
+            for (const item of pricelistItems) {
+                if (!applicableRules[item.pricelist_id.id]) {
+                    applicableRules[item.pricelist_id.id] = [];
+                }
+
+                if (!product.isPricelistItemUsable(item, date)) {
+                    continue;
+                }
+
+                if (item.product_id && product.id === item.product_id.id) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                } else if (
+                    item.raw.product_tmpl_id &&
+                    product.raw?.product_tmpl_id === item.raw.product_tmpl_id
+                ) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                } else if (!item.raw.product_tmpl_id && !item.raw.product_id) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                }
+            }
+
+            product.cachedPricelistRules = applicableRules;
+        }
+    }
+
+    async loadProductPricelist(pricelistIds) {
+        if (!pricelistIds || pricelistIds.length === 0) {
+            return [];
+        }
+
+        const data = await this.data.read("product.pricelist", pricelistIds);
+        const pricelistItemsIds = data.map((p) => p.raw.item_ids).flat();
+
+        if (pricelistItemsIds.length) {
+            await this.data.read("product.pricelist.item", pricelistItemsIds);
+        }
+
+        this.computeProductPricelistCache();
+        return data;
+    }
+
+    async loadProducts(productIds) {
+        if (!productIds || productIds.length === 0) {
+            return [];
+        }
+        this.computeProductPricelistCache();
+        await this.data.read("product.product", productIds);
```

Comment:
```suggestion
        await this.data.read("product.product", productIds);
        this.computeProductPricelistCache();
```
the product should be loaded first before recomputing the pricelist cache.